compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -725,44 +729,72 @@ static void build_chpl_entry_points() {
+  Symbol* opSym = new_CStringSymbol(op);
 
-  FnSymbol* fn = new FnSymbol("==");
+  FnSymbol* fn = new FnSymbol("chpl_fields_are_comparable");
   fn->addFlag(FLAG_COMPILER_GENERATED);
   fn->addFlag(FLAG_LAST_RESORT);
+  fn->addFlag(FLAG_PARAM);
   ArgSymbol* arg1 = new ArgSymbol(INTENT_BLANK, "_arg1", ct);
   arg1->addFlag(FLAG_MARKED_GENERIC);
   ArgSymbol* arg2 = new ArgSymbol(INTENT_BLANK, "_arg2", ct);
   arg2->addFlag(FLAG_MARKED_GENERIC);
   fn->insertFormalAtTail(arg1);
   fn->insertFormalAtTail(arg2);
   fn->retType = dtBool;
-  fn->where = new BlockStmt(new CallExpr("==",
-                                         new CallExpr(PRIM_TYPEOF, arg1),
-                                         new CallExpr(PRIM_TYPEOF, arg2)));
+  fn->retTag = RET_PARAM;
 
-  for_fields(tmp, ct) {
-    if (!tmp->hasFlag(FLAG_IMPLICIT_ALIAS_FIELD)) {
-      Expr* left = new CallExpr(tmp->name, gMethodToken, arg1);
-      Expr* right = new CallExpr(tmp->name, gMethodToken, arg2);
-      fn->insertAtTail(new CondStmt(new CallExpr("!=", left, right), new CallExpr(PRIM_RETURN, gFalse)));
+  if (ct->numFields() == 0) {  // no fields, no need for &&
+    fn->insertAtTail(new CallExpr(PRIM_RETURN, gTrue));
+  }
+  else {  // we have some fields, so possibly && them
+    std::vector<CallExpr *> fieldChecks;
+    for_fields(tmp, ct) {
+      if (!tmp->hasFlag(FLAG_IMPLICIT_ALIAS_FIELD) &&
+          !tmp->hasFlag(FLAG_TYPE_VARIABLE)) {  // types fields must be equal
+        Expr* left = new CallExpr(tmp->name, gMethodToken, arg1);
+        Expr* right = new CallExpr(tmp->name, gMethodToken, arg2);
+        fieldChecks.push_back(new CallExpr(PRIM_CALL_RESOLVES,
+                                           opSym, left, right));
+      }
+    }
+    if (fieldChecks.size() == 0) {
+      fn->insertAtTail(new CallExpr(PRIM_RETURN, gTrue));
+    }
+    else if (fieldChecks.size() == 1) {
+      fn->insertAtTail(new CallExpr(PRIM_RETURN, fieldChecks[0]));
+    }
+    else {
+      CallExpr* boolExpr = new CallExpr(PRIM_AND);
+      for_vector(CallExpr, ce, fieldChecks) {
+        boolExpr->insertAtTail(ce);
+      }
+      fn->insertAtTail(new CallExpr(PRIM_RETURN, boolExpr));
     }
   }
-  fn->insertAtTail(new CallExpr(PRIM_RETURN, gTrue));
+
   DefExpr* def = new DefExpr(fn);
   ct->symbol->defPoint->insertBefore(def);
   reset_ast_loc(def, ct->symbol);
   normalize(fn);
-}
 
+  return fn;
+}
 
-static void build_record_inequality_function(AggregateType* ct) {
-  if (function_exists("!=", ct, ct))
+static void buildRecordComparisonFunc(AggregateType* ct, const char* op) {
+  if (functionExists(op, ct, ct))
     return;
 
-  FnSymbol* fn = new FnSymbol("!=");
+  const char* astrOp = astr(op);
+
+  // we need to special case `!=`:
+  // it can return true early, and returns false after checking all fields
+  // all other operators does the exact opposite
+  bool isNotEqual = (astrOp == astrSne);
+
+  FnSymbol* fn = new FnSymbol(op);
```

Comment:
Should you use the astr for the FnSymbol's name?  I can never remember how astr works ...